File: library/src/test/java/io/github/anderscheow/validator/rules/common/EqualRuleTest.java
package io.github.anderscheow.validator.rules.common;

import android.support.annotation.StringRes;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.util.Locale;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.*;

public class EqualRuleTest {

    private static final String KEYWORD = "test";

    private EqualRule equalRule;

    @Before
    public void setUp() throws Exception {
    }

    @After
    public void tearDown() throws Exception {
        equalRule = null;
    }

    @Test(expected = ClassCastException.class)
    public void validate_NotStringSample_ThrowClassCastException() throws Exception {
        equalRule = new EqualRule(KEYWORD);

        // Any value other than String
        equalRule.validate(123);
    }

    @Test
    public void validate_ValidSample_ReturnTrue() throws Exception {
        equalRule = new EqualRule(KEYWORD);

        String sample = "test";

        assertTrue(equalRule.validate(sample));
    }

    @Test
    public void validate_InvalidSample_ReturnFalse() throws Exception {
        equalRule = new EqualRule(KEYWORD);

        String sample = "TEST";

        assertFalse(equalRule.validate(sample));
    }

    @Test
    public void errorMessage_DefaultErrorMessage() throws Exception {
        String errorMessage = String.format(Locale.getDefault(), "Value does not equal to '%s'", KEYWORD);

        equalRule = new EqualRule(KEYWORD);

        assertEquals(errorMessage, equalRule.errorMessage());
    }

    @Test
    public void errorMessage_CustomErrorMessage() throws Exception {
        String errorMessage = "This is a custom error message";

        equalRule = new EqualRule(KEYWORD, errorMessage);

        assertEquals(errorMessage, equalRule.errorMessage());
    }

    @Test
    public void errorMessage_CustomErrorRes() throws Exception {
        @StringRes int errorRes = 0;

        equalRule = new EqualRule(KEYWORD, errorRes);

        assertEquals(errorRes, equalRule.errorRes());
    }
}

File: library/src/test/java/io/github/anderscheow/validator/rules/common/EqualRuleTest.kt
package io.github.anderscheow.validator.rules.common

import androidx.annotation.StringRes
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import java.util.*

class EqualRuleTest {

    private lateinit var equalRule: EqualRule

    @Before
    @Throws(Exception::class)
    fun setUp() {
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
    }

    @Test(expected = NullPointerException::class)
    @Throws(Exception::class)
    fun validate_EmptySample_ThrowNullPointerException() {
        equalRule = EqualRule(STRING_KEYWORD)

        equalRule.validate(null)
    }

    @Test
    @Throws(Exception::class)
    fun validate_StringSample_ReturnTrue() {
        equalRule = EqualRule(STRING_KEYWORD)

        val sample = "test"

        assertTrue(equalRule.validate(sample))
    }

    @Test
    @Throws(Exception::class)
    fun validate_StringSample_ReturnFalse() {
        equalRule = EqualRule(STRING_KEYWORD)

        val sample = "TEST"

        assertFalse(equalRule.validate(sample))
    }

    @Test
    @Throws(Exception::class)
    fun errorMessage_DefaultErrorMessage() {
        val errorMessage = String.format(Locale.getDefault(), "Value does not equal to '%s'", STRING_KEYWORD)

        equalRule = EqualRule(STRING_KEYWORD)

        assertEquals(errorMessage, equalRule.errorString)
    }

    @Test
    @Throws(Exception::class)
    fun errorMessage_CustomErrorMessage() {
        val errorMessage = "This is a custom error message"

        equalRule = EqualRule(STRING_KEYWORD, errorMessage)

        assertEquals(errorMessage, equalRule.errorString)
    }

    @Test
    @Throws(Exception::class)
    fun errorMessage_CustomErrorRes() {
        @StringRes val errorRes = 0

        equalRule = EqualRule(STRING_KEYWORD, errorRes)

        assertEquals(errorRes, equalRule.errorRes)
    }

    companion object {

        private const val STRING_KEYWORD = "test"
        private const val INT_KEYWORD = 456
        private const val BOOL_KEYWORD = false
    }
}